File: components/applications.py
import os
import re
from PyQt6.QtWidgets import (
    QLabel, QTextEdit, QVBoxLayout, QScrollArea, QWidget, QPushButton,
    QHBoxLayout, QLineEdit, QDialog, QDialogButtonBox, QFormLayout,
    QListWidget, QMessageBox, QCheckBox
)
from PyQt6.QtGui import QFont
from PyQt6.QtCore import Qt, QTimer


class ApplicationsManager:
    def __init__(self, parent):
        self.parent = parent
        self.init_ui()

        # Auto-refresh every 1000ms
        self.timer = QTimer()
        self.timer.timeout.connect(self.refresh_data)
        self.timer.start(5000)

    def init_ui(self):
        self.container = QWidget(self.parent)
        self.container.setGeometry(80, 80, 1440, 640)

        self.container.setStyleSheet("""
            background-color: rgba(255, 255, 255, 0.05);
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
        """)
        self.container.setWindowFlag(Qt.WindowType.WindowStaysOnTopHint, True)
        self.container.setVisible(False)

        layout = QVBoxLayout(self.container)
        layout.setContentsMargins(40, 40, 40, 40)
        layout.setSpacing(40)

        self.text_font = QFont("monospace", 10)
        self.text_font.setStyleHint(QFont.StyleHint.Monospace)

        button_bar = QHBoxLayout()
        add_btn = QPushButton("+ APP")
        remove_btn = QPushButton("- APP")
        add_keybind_btn = QPushButton("+ KB")
        remove_keybind_btn = QPushButton("- KB")

        for btn in (add_btn, remove_btn, add_keybind_btn, remove_keybind_btn):
            btn.setFixedSize(70, 60)
            btn.setStyleSheet("""
                QPushButton {
                    color: white;
                    background-color: rgba(255, 255, 255, 0.05);
                    border: 2px solid rgba(255,255,255,0.3);
                    border-radius: 18px;
                    font-size: 18px;
                    font-weight: bold;
                }
                QPushButton:hover {
                    background-color: rgba(110, 110, 110, 1);
                }
            """)

        add_btn.clicked.connect(self.open_add_dialog)
        remove_btn.clicked.connect(self.open_remove_dialog)
        add_keybind_btn.clicked.connect(self.open_add_keybind_dialog)
        remove_keybind_btn.clicked.connect(self.open_remove_keybind_dialog)

        button_bar.addWidget(add_btn)
        button_bar.addWidget(remove_btn)
        button_bar.addSpacing(20)
        button_bar.addWidget(add_keybind_btn)
        button_bar.addWidget(remove_keybind_btn)
        button_bar.addStretch()
        layout.addLayout(button_bar)

        self.applications_box = self.create_scroll_area("Applications", self.load_applications())
        self.keybinds_box = self.create_scroll_area("Keybinds", self.load_keybinds())

        layout.addWidget(self.applications_box)
        layout.addWidget(self.keybinds_box)

    def refresh_data(self):
        self.applications_box.text_edit.setText(self.load_applications())
        self.keybinds_box.text_edit.setText(self.load_keybinds())

    def create_scroll_area(self, title, content):
        label = QLabel(f"{title.upper()}")
        label.setFont(QFont("ElysiaOSNew", 16, QFont.Weight.Bold))
        label.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")
        label.setGeometry(650, 45, 300, 635)
        label.setAlignment(Qt.AlignmentFlag.AlignCenter)

        text_edit = QTextEdit()
        text_edit.setFont(self.text_font)
        text_edit.setText(content)
        text_edit.setReadOnly(True)
        text_edit.setStyleSheet("color: white; background: transparent; border: none; margin: 0px; padding: 0px")

        scroll = QScrollArea()
        scroll.setWidgetResizable(True)
        container = QWidget()
        vbox = QVBoxLayout(container)
        vbox.setContentsMargins(0, 0, 0, 0)
        vbox.setSpacing(5)
        vbox.addWidget(label)
        vbox.addWidget(text_edit)
        scroll.setWidget(container)
        scroll.setStyleSheet("QScrollArea { background-color: transparent; border: none; }")

        scroll.text_edit = text_edit
        return scroll

    def show(self):
        self.container.setVisible(True)
        self.parent.hide_tiles()
        self.parent.show_back_button()
        self.parent.current_view = "applications"

    def hide(self):
        self.container.setVisible(False)
        self.parent.show_tiles()
        self.parent.hide_back_button()

    def load_applications(self):
        path = os.path.expanduser("~/.config/hypr/applications.conf")
        if not os.path.exists(path):
            return "No applications.conf found."

        self.app_lines = []
        display_lines = []
        with open(path, "r") as f:
            for line in f:
                stripped = line.strip()
                self.app_lines.append(line.rstrip("\n"))
                if stripped.startswith("#") or not stripped:
                    continue
                if "=" in stripped:
                    key, cmd = map(str.strip, stripped.split("=", 1))
                    key_name = key.replace("$", "").capitalize()
                    display_lines.append(f"{key_name:<30} {cmd}")
        return "\n".join(display_lines[:30])

    def load_keybinds(self):
        path = os.path.expanduser("~/.config/hypr/app_keybinds.conf")
        if not os.path.exists(path):
            return "No app_keybinds.conf found."

        variables = {}
        binds = []

        with open(path, "r") as f:
            for line in f:
                line = line.strip()
                if line.startswith("$") and "=" in line:
                    var, val = map(str.strip, line.split("=", 1))
                    variables[var] = val

        bind_count = 1
        with open(path, "r") as f:
            for line in f:
                original = line.strip()
                if not original.startswith("bind"):
                    continue

                line = re.sub(r"#.*", "", original).strip()
                if "=" in line:
                    line = line.split("=", 1)[1].strip()

                parts = [p.strip() for p in line.split(",")]
                if len(parts) >= 3:
                    mod, key, action_type = parts[:3]
                    action = ",".join(parts[3:]).strip() if len(parts) > 3 else ""

                    for var, val in variables.items():
                        mod = mod.replace(var, val)
                        action = action.replace(var, val)

                    if action_type == "exec":
                        keybind_str = f"BIND {bind_count:<3} {mod} KEY + {key:<6} {action}"
                    else:
                        keybind_str = f"BIND {bind_count:<3} {mod} KEY + {key:<6} {action_type}"
                    binds.append(keybind_str)
                    bind_count += 1

        return "\n".join(binds) if binds else "No keybinds found."

    # --- Remaining dialog code is unchanged (add/remove application/keybinds) ---
    # It's all retained from your original code, ensuring full functionality.
    # If needed, I can re-include that as well.

    # Keep using self.refresh_data() after any edit operations


    def open_add_dialog(self):
        dialog = QDialog(self.container)
        dialog.setWindowTitle("Add Application")
        dialog.setFixedSize(400, 180)
        dialog.setStyleSheet("""
            background-color: #202020;
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
            color: white;
            font-family: 'ElysiaOSNew';
            font-size: 14px;
        """)

        form = QFormLayout(dialog)
        app_name_input = QLineEdit()
        app_exec_input = QLineEdit()
        form.addRow("Application Name:", app_name_input)
        form.addRow("Application Exec:", app_exec_input)

        buttons = QDialogButtonBox(QDialogButtonBox.StandardButton.Ok | QDialogButtonBox.StandardButton.Cancel)
        buttons.accepted.connect(lambda: self.add_application(dialog, app_name_input.text(), app_exec_input.text()))
        buttons.rejected.connect(dialog.reject)
        form.addRow(buttons)

        dialog.exec()

    def add_application(self, dialog, name, exec_path):
        name = name.strip()
        exec_path = exec_path.strip()

        if not name or not exec_path:
            QMessageBox.warning(dialog, "Input Error", "Both Application Name and Exec path are required.")
            return

        conf_path = os.path.expanduser("~/.config/hypr/applications.conf")
        with open(conf_path, "a") as f:
            f.write(f"\n${name} = {exec_path}\n")

        updated_content = self.load_applications()
        self.applications_box.text_edit.setText(updated_content)
        dialog.accept()

    def open_remove_dialog(self):
        dialog = QDialog(self.container)
        dialog.setWindowTitle("Remove Application")
        dialog.setFixedSize(500, 400)
        dialog.setStyleSheet("""
            background-color: #202020;
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
            color: white;
            font-family: 'ElysiaOSNew', monospace;
            font-size: 14px;
        """)

        vbox = QVBoxLayout(dialog)
        list_widget = QListWidget()
        deletable = []

        for line in self.app_lines:
            if line.strip().startswith("$") and "=" in line:
                deletable.append(line)
                key, cmd = map(str.strip, line.split("=", 1))
                key_name = key.replace("$", "").capitalize()
                list_widget.addItem(f"{key_name:<30} {cmd}")

        vbox.addWidget(list_widget)

        buttons = QDialogButtonBox()
        apply_btn = buttons.addButton("Apply Changes", QDialogButtonBox.ButtonRole.AcceptRole)
        cancel_btn = buttons.addButton("Cancel", QDialogButtonBox.ButtonRole.RejectRole)
        vbox.addWidget(buttons)

        def apply_removal():
            selected = list_widget.currentRow()
            if selected < 0:
                QMessageBox.warning(dialog, "No selection", "Please select an entry to remove.")
                return

            to_remove = deletable[selected]
            self.app_lines.remove(to_remove)

            path = os.path.expanduser("~/.config/hypr/applications.conf")
            with open(path, "w") as f:
                for line in self.app_lines:
                    f.write(line + "\n")

            updated_content = self.load_applications()
            self.applications_box.text_edit.setText(updated_content)
            dialog.accept()

        apply_btn.clicked.connect(apply_removal)
        cancel_btn.clicked.connect(dialog.reject)
        dialog.exec()

    # Keybind dialogs and functions

    def open_add_keybind_dialog(self):
        dialog = QDialog(self.container)
        dialog.setWindowTitle("Add Keybind")
        dialog.setFixedSize(400, 260)
        dialog.setStyleSheet("""
            background-color: #202020;
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
            color: white;
            font-family: 'ElysiaOSNew', monospace;
            font-size: 14px;
        """)

        form = QFormLayout(dialog)
        self.keybind1_input = QLineEdit()
        self.keybind2_input = QLineEdit()
        self.keybind3_input = QLineEdit()
        self.exec_checkbox = QCheckBox("Exec")
        self.application_call_input = QLineEdit()

        form.addRow("Keybind1 (e.g. $mainMod):", self.keybind1_input)
        form.addRow("Keybind2 (e.g. L):", self.keybind2_input)
        form.addRow("Keybind3 (optional):", self.keybind3_input)
        form.addRow(self.exec_checkbox)
        form.addRow("Application Call:", self.application_call_input)

        buttons = QDialogButtonBox(QDialogButtonBox.StandardButton.Ok | QDialogButtonBox.StandardButton.Cancel)
        buttons.accepted.connect(lambda: self.add_keybind(dialog))
        buttons.rejected.connect(dialog.reject)
        form.addRow(buttons)

        dialog.exec()

    def add_keybind(self, dialog):
        kb1 = self.keybind1_input.text().strip()
        kb2 = self.keybind2_input.text().strip()
        kb3 = self.keybind3_input.text().strip()
        is_exec = self.exec_checkbox.isChecked()
        app_call = self.application_call_input.text().strip()

        if not kb1 or not kb2 or not app_call:
            QMessageBox.warning(dialog, "Input Error", "Keybind1, Keybind2, and Application Call are required.")
            return

        parts = [kb1, kb2]
        if kb3:
            parts.append(kb3)

        if is_exec:
            bind_line = f"bind = {', '.join(parts)}, exec, {app_call}"
        else:
            bind_line = f"bind = {', '.join(parts)}, {app_call}"

        conf_path = os.path.expanduser("~/.config/hypr/app_keybinds.conf")
        with open(conf_path, "a") as f:
            f.write("\n" + bind_line + "\n")

        updated_content = self.load_keybinds()
        self.keybinds_box.text_edit.setText(updated_content)
        dialog.accept()

    def open_remove_keybind_dialog(self):
        dialog = QDialog(self.container)
        dialog.setWindowTitle("Remove Keybind")
        dialog.setFixedSize(600, 400)
        dialog.setStyleSheet("""
            background-color: #202020;
            border: 2px solid rgba(255, 255, 255, 0.15);
            border-radius: 30px;
            color: white;
            font-family: 'ElysiaOSNew', monospace;
            font-size: 14px;
        """)

        vbox = QVBoxLayout(dialog)
        list_widget = QListWidget()
        deletable = []

        path = os.path.expanduser("~/.config/hypr/app_keybinds.conf")
        if not os.path.exists(path):
            QMessageBox.information(dialog, "Info", "No keybind config found.")
            dialog.reject()
            return

        with open(path, "r") as f:
            for line in f:
                line_strip = line.strip()
                if line_strip.startswith("bind"):
                    deletable.append(line_strip)
                    list_widget.addItem(line_strip)

        vbox.addWidget(list_widget)

        buttons = QDialogButtonBox()
        apply_btn = buttons.addButton("Apply Changes", QDialogButtonBox.ButtonRole.AcceptRole)
        cancel_btn = buttons.addButton("Cancel", QDialogButtonBox.ButtonRole.RejectRole)
        vbox.addWidget(buttons)

        def apply_removal():
            selected = list_widget.currentRow()
            if selected < 0:
                QMessageBox.warning(dialog, "No selection", "Please select a keybind to remove.")
                return

            to_remove = deletable[selected]
            deletable.pop(selected)

            with open(path, "r") as f_read:
                lines = f_read.readlines()

            with open(path, "w") as f_write:
                for line in lines:
                    if line.strip() != to_remove:
                        f_write.write(line)

            updated_content = self.load_keybinds()
            self.keybinds_box.text_edit.setText(updated_content)
            dialog.accept()

        apply_btn.clicked.connect(apply_removal)
        cancel_btn.clicked.connect(dialog.reject)
        dialog.exec()

File: components/Applications.cpp
#include "Applications.h"
#include "../translations.h"
#include <QDir>
#include <QFont>
#include <QProcess>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QScrollArea>
#include <QTextEdit>
#include <QDialog>
#include <QDialogButtonBox>
#include <QFormLayout>
#include <QLineEdit>
#include <QListWidget>
#include <QMessageBox>
#include <QTimer>
#include <QFile>
#include <QTextStream>
#include <QRegularExpression>
#include <QCheckBox>
#include <QFileSystemWatcher>

Applications::Applications(QObject *parent) : QObject(parent), currentConfigIdx(0) {
    configFiles = { 
        QDir::homePath() + "/.config/hypr/applications.conf",
        QDir::homePath() + "/.config/hypr/app_keybinds.conf",
        QDir::homePath() + "/.config/hypr/auto_start.conf",
        QDir::homePath() + "/.config/hypr/window_rules.conf"
    };
    configNames = { Translations::tr("APPLICATIONS"), Translations::tr("KEYBINDS"), Translations::tr("AUTORUNS"), Translations::tr("WINDOW_RULES") };
    QWidget *parentWidget = qobject_cast<QWidget*>(parent);
    if (!parentWidget) return;
    initUI();
    fileWatcher = new QFileSystemWatcher(this);
    fileWatcher->addPath(configFiles[0]);
    fileWatcher->addPath(configFiles[1]);
    fileWatcher->addPath(configFiles[2]);
    fileWatcher->addPath(configFiles[3]);
    connect(fileWatcher, &QFileSystemWatcher::fileChanged, this, [this](const QString &path) {
        loadConfig();
        fileWatcher->addPath(path); // Re-add in case it was removed after change
    });
}

void Applications::initUI() {
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (!parentWidget) return;
    container = new QWidget(parentWidget);
    container->setGeometry(80, 80, 1440, 640);
    container->setStyleSheet(
        "background-color: rgba(255, 255, 255, 0.1);"
        "border: 36px;"
        "border-radius: 36px;"
        "margin: 4px;"
    );
    container->setVisible(false);
    mainLayout = new QVBoxLayout(container);
    mainLayout->setContentsMargins(50, 50, 50, 50);
    mainLayout->setSpacing(40);
    // Header with arrows and config name
    headerWidget = new QWidget(container);
    QHBoxLayout *headerLayout = new QHBoxLayout(headerWidget);
    headerLayout->setContentsMargins(0, 0, 0, 0);
    leftArrow = new QPushButton("<", headerWidget);
    rightArrow = new QPushButton(">", headerWidget);
    leftArrow->setFixedSize(50, 50);
    rightArrow->setFixedSize(50, 50);
    leftArrow->setStyleSheet("QPushButton { color: rgba(255,255,255,0.8); background: transparent; border: none; font-size: 24px; font-weight: bold; } QPushButton:hover { color: #e5a7c6; }");
    rightArrow->setStyleSheet("QPushButton { color: rgba(255,255,255,0.8); background: transparent; border: none; font-size: 24px; font-weight: bold; } QPushButton:hover { color: #e5a7c6; }");
    configLabel = new QLabel(configNames[currentConfigIdx], headerWidget);
    configLabel->setFont(QFont("ElysiaOSNew", 20, QFont::Bold));
    configLabel->setStyleSheet("color: rgba(255,255,255,0.95); background: transparent; border: none; border-radius: 0px;");
    configLabel->setAlignment(Qt::AlignCenter);
    headerLayout->addWidget(leftArrow);
    headerLayout->addStretch();
    headerLayout->addWidget(configLabel);
    headerLayout->addStretch();
    headerLayout->addWidget(rightArrow);
    mainLayout->addWidget(headerWidget);
    connect(leftArrow, &QPushButton::clicked, this, &Applications::handleLeftArrow);
    connect(rightArrow, &QPushButton::clicked, this, &Applications::handleRightArrow);
    // Scroll area for lines
    QScrollArea *scrollArea = new QScrollArea(container);
    scrollArea->setWidgetResizable(true);
    scrollArea->setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
    scrollArea->setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
    scrollArea->setStyleSheet("QScrollArea { background: transparent; border: none; } QScrollBar:vertical { background: transparent; border-radius: 6px; width: 12px; } QScrollBar::handle:vertical { background: rgba(229,167,198,0.5); border-radius: 6px; } QScrollBar::handle:vertical:hover { background: rgba(229,167,198,0.8); }");
    linesWidget = new QWidget(scrollArea);
    linesWidget->setStyleSheet(
        "background-color: transparent;"
        "border: 0px;"
        "border-radius: 36px;"
        "margin: 0px;"
    );
    linesLayout = new QVBoxLayout(linesWidget);
    linesLayout->setContentsMargins(20, 20, 20, 20);
    linesLayout->setSpacing(20);
    scrollArea->setWidget(linesWidget);
    mainLayout->addWidget(scrollArea);
    loadConfig();
}

void Applications::loadConfig() {
    QStringList allLines;
    QFile file(configFiles[currentConfigIdx]);
    if (file.exists() && file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        QTextStream in(&file);
        while (!in.atEnd()) allLines.append(in.readLine());
        file.close();
    }
    // Separate header based on current config type
    int firstConfigIdx = 0;
    if (currentConfigIdx == 0) {
        // Applications config - ignore ########, ##APPS##, ########
        for (int i = 0; i < allLines.size(); ++i) {
            QString line = allLines[i].trimmed();
            if (line.startsWith("########") || line.startsWith("##APPS##")) {
                firstConfigIdx = i + 1;
            }
        }
    } else if (currentConfigIdx == 1) {
        // Keybinds config - ignore ########################, ########KEYBINDS########, ########################, and the comment line
        for (int i = 0; i < allLines.size(); ++i) {
            QString line = allLines[i].trimmed();
            if (line.startsWith("########################") || line.startsWith("########KEYBINDS########")) {
                firstConfigIdx = i + 1;
            }
        }
        // Skip the comment line after header
        if (firstConfigIdx < allLines.size() && allLines[firstConfigIdx].trimmed().startsWith("# See https://wiki.hyprland.org/Configuring/Keywords/ for more")) {
            firstConfigIdx++;
        }
    } else if (currentConfigIdx == 2) {
        // Autoruns config - ignore ########################, ########AUTORUNS########, ########################, and the comment line
        for (int i = 0; i < allLines.size(); ++i) {
            QString line = allLines[i].trimmed();
            if (line.startsWith("########################") || line.startsWith("########AUTORUNS########")) {
                firstConfigIdx = i + 1;
            }
        }
        // Skip the comment line after header
        if (firstConfigIdx < allLines.size() && allLines[firstConfigIdx].trimmed().startsWith("# Execute your favorite apps at launch")) {
            firstConfigIdx++;
        }
    } else {
        // Window Rules config - ignore ################################, ##TRANSPARENT BACKGROUND APP####, ################################, and all the example/comment lines
        for (int i = 0; i < allLines.size(); ++i) {
            QString line = allLines[i].trimmed();
            if (line.startsWith("################################") || line.startsWith("##TRANSPARENT BACKGROUND APP####")) {
                firstConfigIdx = i + 1;
            }
        }
        // Skip all the example and comment lines after header
        while (firstConfigIdx < allLines.size()) {
            QString line = allLines[firstConfigIdx].trimmed();
            if (line.startsWith("#") || line.isEmpty()) {
                firstConfigIdx++;
            } else {
                break;
            }
        }
    }
    preservedHeader = allLines.mid(0, firstConfigIdx);
    currentLines.clear();
    for (int i = firstConfigIdx; i < allLines.size(); ++i) {
        currentLines.append(allLines[i]);
    }
    rebuildLines();
}

QStringList Applications::readConfigLines(const QString &path) {
    QStringList result;
    QFile file(path);
    if (!file.exists() || !file.open(QIODevice::ReadOnly | QIODevice::Text)) return result;
    QTextStream in(&file);
    while (!in.atEnd()) {
        QString line = in.readLine().trimmed();
        if (line.isEmpty() || line.startsWith("#")) continue;
        if (line.contains("=")) result.append(line);
    }
    return result;
}

void Applications::rebuildLines() {
    // Clear old widgets
    QLayoutItem *child;
    while ((child = linesLayout->takeAt(0)) != nullptr) {
        if (child->widget()) child->widget()->deleteLater();
        delete child;
    }
    lineEdits.clear();
    for (int i = 0; i < currentLines.size(); ++i) {
        QWidget *row = new QWidget(linesWidget);
        QHBoxLayout *rowLayout = new QHBoxLayout(row);
        rowLayout->setContentsMargins(0, 0, 0, 0);
        rowLayout->setSpacing(20);
        QLineEdit *edit = new QLineEdit(currentLines[i], row);
        edit->setFont(QFont("monospace", 14));
        edit->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
        edit->setStyleSheet("QLineEdit { color: white; background: transparent; border: none; margin: 0px; padding: 0px; outline: none; } QLineEdit:focus { color: white; background: transparent; border: none; margin: 0px; padding: 0px; outline: none; } QLineEdit:hover { color: white; background: transparent; border: none; margin: 0px; padding: 0px; }");
        connect(edit, &QLineEdit::editingFinished, this, [this, i, edit]() { handleLineEdited(i); });
        QPushButton *addBtn = new QPushButton("+", row);
        QPushButton *removeBtn = new QPushButton("-", row);
        addBtn->setFixedSize(50, 50);
        removeBtn->setFixedSize(50, 50);
        addBtn->setStyleSheet("QPushButton { color: #e5a7c6; background: rgba(255,255,255,0.8); border: 1px solid rgba(229,167,198,0.6); border-radius: 8px; font-size: 18px; font-weight: bold; } QPushButton:hover { background: rgba(229,167,198,0.15); color: white; }");
        removeBtn->setStyleSheet("QPushButton { color: #e57373; background: rgba(255,255,255,0.8); border: 1px solid rgba(229,103,115,0.6); border-radius: 8px; font-size: 18px; font-weight: bold; } QPushButton:hover { background: rgba(229,103,115,0.15); color: white; }");
        connect(addBtn, &QPushButton::clicked, this, [this, i]() { handleAddLine(i); });
        connect(removeBtn, &QPushButton::clicked, this, [this, i]() { handleRemoveLine(i); });
        rowLayout->addWidget(edit, 1);
        rowLayout->addWidget(addBtn);
        rowLayout->addWidget(removeBtn);
        row->setLayout(rowLayout);
        linesLayout->addWidget(row);
        lineEdits.append(edit);
    }
    linesLayout->addStretch();
}

void Applications::handleLineEdited(int idx) {
    if (idx < 0 || idx >= lineEdits.size()) return;
    currentLines[idx] = lineEdits[idx]->text();
    saveConfig();
}

void Applications::handleAddLine(int idx) {
    if (idx < 0 || idx > currentLines.size()) return;
    currentLines.insert(idx + 1, "");
    saveConfig();
    rebuildLines();
}

void Applications::handleRemoveLine(int idx) {
    if (idx < 0 || idx >= currentLines.size()) return;
    currentLines.removeAt(idx);
    saveConfig();
    rebuildLines();
}

void Applications::saveConfig() {
    QFile file(configFiles[currentConfigIdx]);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) return;
    QTextStream out(&file);
    for (const QString &line : preservedHeader) out << line << "\n";
    for (const QString &line : currentLines) out << line << "\n";
    file.close();
    // Re-add watcher in case file was recreated
    if (!fileWatcher->files().contains(configFiles[currentConfigIdx]))
        fileWatcher->addPath(configFiles[currentConfigIdx]);
}

void Applications::writeConfigLines(const QString &path, const QStringList &lines) {
    QFile file(path);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text | QIODevice::Truncate)) return;
    QTextStream out(&file);
    for (const QString &line : lines) out << line << "\n";
    file.close();
}

void Applications::refreshData() {
    loadConfig();
}

void Applications::handleLeftArrow() {
    currentConfigIdx = (currentConfigIdx - 1 + configFiles.size()) % configFiles.size();
    configLabel->setText(configNames[currentConfigIdx]);
    loadConfig();
}

void Applications::handleRightArrow() {
    currentConfigIdx = (currentConfigIdx + 1) % configFiles.size();
    configLabel->setText(configNames[currentConfigIdx]);
    loadConfig();
}
void Applications::show() {
    container->setVisible(true);
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("hideTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideTiles");
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("showBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showBackButton");
}
void Applications::hide() {
    container->setVisible(false);
    QWidget *parentWidget = qobject_cast<QWidget*>(parent());
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("showTiles()") != -1)
        QMetaObject::invokeMethod(parentWidget, "showTiles");
    if (parentWidget && parentWidget->metaObject()->indexOfMethod("hideBackButton()") != -1)
        QMetaObject::invokeMethod(parentWidget, "hideBackButton");
} 